ArduPlane/ArduPlane.cpp

```diff
@@ -845,43 +845,117 @@ void Plane::update_flight_stage(void)
+              make is_flying() more accurate during various auto modes
+             */
+
+            switch (flight_stage)
+            {
+            case AP_SpdHgtControl::FLIGHT_TAKEOFF:
+                // while on the ground, an uncalibrated airspeed sensor can drift to 7m/s so
+                // ensure we aren't showing a false positive. If the throttle is suppressed
+                // we are definitely not flying, or at least for not much longer!
+                if (throttle_suppressed) {
+                    is_flying_bool = false;
+                    auto_state.is_crashed = false;
+                }
+                break;
 
+            case AP_SpdHgtControl::FLIGHT_NORMAL:
+                // TODO: detect ground impacts
+                break;
 
-    if (hal.util->get_soft_armed()) {
-        // when armed, we need overwhelming evidence that we ARE NOT flying
-        isFlyingBool = airspeedMovement || gpsMovement;
+            case AP_SpdHgtControl::FLIGHT_LAND_APPROACH:
+                // TODO: detect ground impacts
+                if (fabsf(auto_state.sink_rate) > 0.2f) {
+                    is_flying_bool = true;
+                }
+                break;
 
-        /*
-          make is_flying() more accurate for landing approach
-         */
-        if (flight_stage == AP_SpdHgtControl::FLIGHT_LAND_APPROACH &&
-            fabsf(auto_state.sink_rate) > 0.2f) {
-            isFlyingBool = true;
+            case AP_SpdHgtControl::FLIGHT_LAND_FINAL:
+            default:
+                break;
+            } // switch
         }
     } else {
-        // when disarmed, we need overwhelming evidence that we ARE flying
-        isFlyingBool = airspeedMovement && gpsMovement;
+        // when disarmed assume not flying and need overwhelming evidence that we ARE flying
+
+        is_flying_bool = airspeed_movement && gps_confirmed_movement;
+
+        if ((control_mode == AUTO) &&
+            ((flight_stage == AP_SpdHgtControl::FLIGHT_TAKEOFF) ||
+             (flight_stage == AP_SpdHgtControl::FLIGHT_LAND_FINAL)) ) {
+            is_flying_bool = false;
+        }
     }
 
     // low-pass the result.
-    isFlyingProbability = (0.6f * isFlyingProbability) + (0.4f * (float)isFlyingBool);
+    // coef=0.15f @ 5Hz takes 3.0s to go from 100% down to 10% (or 0% up to 90%)
+    isFlyingProbability = (0.85f * isFlyingProbability) + (0.15f * (float)is_flying_bool);
 
     /*
       update last_flying_ms so we always know how long we have not
       been flying for. This helps for crash detection and auto-disarm
      */
-    if (is_flying()) {
-        auto_state.last_flying_ms = millis();
+    bool new_is_flying = is_flying();
+    static bool previous_is_flying = false;
+
+    // we are flying, note the time
+    if (new_is_flying) {
+
+        auto_state.last_flying_ms = now_ms;
+
+        if ((control_mode == AUTO) &&
+            ((auto_state.started_flying_in_auto_ms == 0) || (new_is_flying != previous_is_flying)) ) {
```

Comment:
It's probably a style thing, and not super relevant because compilers are clever like that but (new_is_flying != previous_is_flying) is the equivalent to (true != previous_is_flying) so just (!previous_is_flying) would work fine there.